lib/handlers/revoke-handler.ts

```diff
@@ -0,0 +1,341 @@
+        'Invalid argument: model does not implement `getClient()`',
+      );
+    }
+
+    if (!options.model.getRefreshToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getRefreshToken()`',
+      );
+    }
+
+    if (!options.model.getAccessToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getAccessToken()`',
+      );
+    }
+
+    if (!options.model.revokeToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `revokeToken()`',
+      );
+    }
+
+    this.model = options.model;
+  }
+
+  /**
+   * Revoke Handler.
+   */
+
+  async handle(request: Request, response: Response) {
+    if (!(request instanceof Request)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `request` must be an instance of Request',
+      );
+    }
+
+    if (!(response instanceof Response)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `response` must be an instance of Response',
+      );
+    }
+
+    if (request.method !== 'POST') {
+      throw new InvalidRequestError('Invalid request: method must be POST');
+    }
+
+    if (!request.is('application/x-www-form-urlencoded')) {
+      throw new InvalidRequestError(
+        'Invalid request: content must be application/x-www-form-urlencoded',
+      );
+    }
+    // Extend model object with request
+    this.model.request = request;
+
+    try {
+      const client = await this.getClient(request, response);
+
+      return this.handleRevokeToken(request, client);
+    } catch (e) {
+      let error = e;
+      if (!(error instanceof OAuthError)) {
+        error = new ServerError(error);
+      }
+      /**
+       * All necessary information is conveyed in the response code.
+       *
+       * Note: invalid tokens do not cause an error response since the client
+       * cannot handle such an error in a reasonable way.  Moreover, the
+       * purpose of the revocation request, invalidating the particular token,
+       * is already achieved.
+       * @see https://tools.ietf.org/html/rfc7009#section-2.2
+       */
+      if (!(error instanceof InvalidTokenError)) {
+        this.updateErrorResponse(response, error);
+      }
+
+      throw error;
+    }
+  }
+
+  /**
+   * Revoke a refresh or access token.
+   *
+   * Handle the revoking of refresh tokens, and access tokens if supported / desirable
+   * RFC7009 specifies that "If the server is unable to locate the token using
+   * the given hint, it MUST extend its search across all of its supported token types"
+   */
+
+  async handleRevokeToken(request: Request, client: Client) {
+    try {
+      let token = await this.getTokenFromRequest(request);
+      token = await oneSuccess([
+        this.getAccessToken(token, client),
+        this.getRefreshToken(token, client),
+      ]);
+
+      return this.revokeToken(token);
+    } catch (errors) {
```

Comment:
Add logic to this catch clause or eliminate it and rethrow the exception automatically